src/lib/y2partitioner/blk_device_restorer.rb

```diff
@@ -0,0 +1,209 @@
+    # @param dev [Y2Storage::BlkDevice, nil] source device
+    # @return [Boolean]
+    def can_restore_device?(dev)
+      return false if dev.nil?
+      return true if dev.descendants.empty?
+      return true if dev.filesystem
+      dev.descendants.size == 1 && dev.encrypted?
+    end
+
+    # Restores a given source device into {#device}
+    #
+    # @see #restore_from_system
+    # @see #restore_from_checkpoint
+    #
+    # @param source [Y2Storage::BlkDevice, nil]
+    def restore_device(source)
+      return unless can_restore_device?(source)
+
+      device.remove_descendants
+      device.id = source.id if source.is?(:partition)
+
+      restore_children(source) if source.has_children?
+    end
+
+    # @param source [Y2Storage::BlkDevice]
+    def restore_children(source)
+      if source.encrypted?
+        restore_encryption_and_descendants(source)
+      else
+        restore_filesystem_and_descendants(source)
+      end
+    end
+
+    # @param source [Y2Storage::BlkDevice]
+    def restore_encryption_and_descendants(source)
+      encryption = source.encryption.to_storage_value
+      copy_device(encryption)
+      restore_filesystem_and_descendants(source) if source.filesystem
+    end
+
+    # @param source [Y2Storage::BlkDevice]
+    def restore_filesystem_and_descendants(source)
+      filesystem = source.filesystem.to_storage_value
+      copy_device(filesystem)
+      copy_subvolumes(filesystem)
+    end
+
+    # Recursively copy all the subvolumes of the source device to the current
+    # devicegraph, connecting each of them to its corresponding parent.
+    #
+    # @raise [RuntimeError] if any subvolume has more than one parent
+    #
+    # @param source [Storage::Device] source device, beware: is a low-level
+    #   Storage device, not a Y2Storage one
+    def copy_subvolumes(source)
+      source.children.each do |child|
+        next unless Storage.btrfs_subvolume?(child)
+        copy_device(child)
+        copy_subvolumes(child)
+      end
+    end
+
+    # Copy the source device to the current devicegraph, connecting it to its
+    # corresponding parent
+    #
+    # @raise [RuntimeError] if the device has more than one parent
+    #
+    # @param source [Storage::Device] source device, beware: is a low-level
+    #   Storage device, not a Y2Storage one
+    def copy_device(source)
+      if source.in_holders.size != 1
+        log.error "The device has more than one parent, that's unexpected: #{source.sid}"
+        raise "Unexpected error restoring the status of device #{source.sid}"
+      end
+      source.copy_to_devicegraph(raw_current_graph)
+      source.in_holders[0].copy_to_devicegraph(raw_current_graph)
+    end
+
+    # Equivalent to {#device} in the system devicegraph, if any
+    #
+    # @return [Y2Storage::BlkDevice, nil]
+    def system_device
+      DeviceGraphs.instance.system.find_device(device.sid)
+    end
+
+    # Equivalent to {#device} in its checkpoint devicegraph, if any
+    #
```

Comment:
NP: I would explain that the device is searched into system when there is no checkpoint, and it returns nil when the device does not exist (nor in system).